rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py

```diff
@@ -0,0 +1,924 @@
+        # Map the index in the trees to the index in `self._indices`.
+        self._module_to_tree_idx_to_sample_idx[module_id][idx] = sample_index
+        # Return the index.
+        return idx
+
+    @override(MultiAgentEpisodeReplayBuffer)
+    def _sample_independent(
+        self,
+        batch_size_B: Optional[int],
+        batch_length_T: Optional[int],
+        n_step: Optional[Union[int, Tuple]],
+        gamma: float,
+        include_infos: bool,
+        include_extra_model_outputs: bool,
+        modules_to_sample: Optional[List[ModuleID]],
+        beta: Optional[float],
+    ) -> List["SingleAgentEpisode"]:
+        """Samples a single-agent episode list with independent transitions.
+
+        Note, independent sampling means that each module samples its transitions
+        independently from the replay buffer. This is the default sampling mode.
+        In contrast, synchronized sampling samples transitions from the same
+        environment step.
+        """
+
+        actual_n_step = n_step or 1
+        # Sample the n-step if necessary.
+        random_n_step = isinstance(n_step, tuple)
+
+        # Keep track of the indices that were sampled last for updating the
+        # weights later (see `ray.rllib.utils.replay_buffer.utils.
+        # update_priorities_in_episode_replay_buffer`).
+        # self._last_sampled_indices = defaultdict(lambda: [])
+
+        sampled_episodes = []
+        # TODO (simon): Ensure that the module has data and if not, skip it.
+        # TODO (sven): Should we then error out or skip? I think the Learner
+        # should handle this case when a module has no train data.
+        modules_to_sample = modules_to_sample or set(self._module_to_indices.keys())
+        for module_id in modules_to_sample:
+            # Sample proportionally from the replay buffer's module segments using the
+            # respective weights.
+            module_total_segment_sum = self._module_to_sum_segment[module_id].sum()
+            module_p_min = (
+                self._module_to_min_segment[module_id].min() / module_total_segment_sum
+            )
+            # TODO (simon): Allow individual betas per module.
+            module_max_weight = (module_p_min * self.get_num_timesteps(module_id)) ** (
+                -beta
+            )
+            B = 0
+            while B < batch_size_B:
+                # First, draw a random sample from Uniform(0, sum over all weights).
+                # Note, transitions with higher weight get sampled more often (as
+                # more random draws fall into larger intervals).
+                module_random_sum = (
+                    self.rng.random() * self._module_to_sum_segment[module_id].sum()
+                )
+                # Get the highest index in the sum-tree for which the sum is
+                # smaller or equal the random sum sample.
+                # Note, we sample `o_(t + n_step)` as this is the state that
```

Comment:
Should we better say: "We sample the transition `o_(t)` to `o_(t + n_step)` as this ..."

Reply:
I changed this comment completely as it is out-dated: "Note, in contrast to Schaul et al. (2018) (who sample `o_(t + n_step)`, Algorithm 1) we sample `o_t`."